x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java

```diff
@@ -339,25 +370,90 @@ ResolvedIndices resolveIndicesAndAliases(
                 // if we cannot replace wildcards the indices list stays empty. Same if there are no authorized indices.
                 // we honour allow_no_indices like es core does.
             } else {
+                if (indicesRequest instanceof IndicesRequest.CrossProjectReplaceable crossProjectReplaceableRequest
+                    && targetProjects != AuthorizedProjectsSupplier.AuthorizedProjects.NOT_CROSS_PROJECT) {
+                    assert crossProjectReplaceableRequest.allowsRemoteIndices();
+                    // TODO ideally, we'd consolidate this with `resolveIndexAbstractionsCrossProject`
+                    CrossProjectResolverUtils.maybeRewriteCrossProjectResolvableRequest(
+                        remoteClusterResolver,
+                        targetProjects,
+                        crossProjectReplaceableRequest
+                    );
+                    logger.info(
+                        "Cross-project search applied, indices are now [{}]",
+                        Arrays.toString(crossProjectReplaceableRequest.indices())
+                    );
+                    if (crossProjectReplaceableRequest.shouldApplyCrossProjectHandling()) {
+                        logger.info("Handling as CPS request for [{}]", Arrays.toString(crossProjectReplaceableRequest.indices()));
+                        // we can re-use resolveIndexAbstractionsMapping() for local indices here
+                        Map<String, IndicesRequest.ReplacedExpression> replacedExpressions = indexAbstractionResolver
+                            .resolveIndexAbstractionsCrossProject(
+                                crossProjectReplaceableRequest.getReplacedExpressions(),
+                                indicesOptions,
+                                projectMetadata,
+                                authorizedIndices::all,
+                                authorizedIndices::check,
+                                indicesRequest.includeDataStreams()
+                            );
+                        crossProjectReplaceableRequest.setReplacedExpressions(replacedExpressions);
+                        // TODO handle empty case by calling
+                        // replaceable.indices(IndicesAndAliasesResolverField.NO_INDICES_OR_ALIASES_ARRAY);
+                        return remoteClusterResolver.splitLocalAndRemoteIndexNames(crossProjectReplaceableRequest.indices());
+                    }
+                }
+
+                logger.info(
+                    "Resolving indices for request [{}] and [{}]",
+                    Arrays.toString(replaceable.indices()),
+                    replaceable.getReplacedExpressions()
+                );
+
+                // TODO why on earth would this happen?
```

Comment:
This is totally bizarre but probably just an artifact of the POC being messy, rather than a fundamental issue with the approac: somehow, when we have a request that results in indices getting replaced with `*.-*` the `resolve()` method is called _twice_ for the same action, and request. It's very tempting to go down the rabbit hole of finding the race condition here but I'm resisting it for now.